### PR TITLE
SBERDOMA-1122 strong GraphQL type for `Property.map` JSON field

### PR DESCRIPTION
**Pull request by AntonAL**
Changed JSON-fields into GraphQL types in following places:

- [x] `Property.map`
- [x] `Property.addressMeta`
- [x] `addressMeta` input variable of `registerResident` mutation
- [x] `addressMeta` input variable of `checkPropertyWithAddressExist` custom query

Added missing fields to factories
Values of some fields was taken from official sources
Initially, I've tried to make strong GraphQL type for `addressMeta.data.house_type` fields, but did't found, how to define custom values for enum items, so, left type `String` for it.

Since adding GraphQL-types instead of untyped JSON for `Property.map` field, a data flow between form and GraphQL API has changed.
Now we are explicitly requesting all subfields of `Property.map` in our queries and get following side-effects:

1. When a field is missing in data, we get `null` value in GraphQL response. It means, when in database following object is stored `{ a: 1 }` and we are requesting `{ a b }`, then we get `{ a: 1, b: null }`
2. For each typed node we get `__typename` field

All these side-effects resulting to wrong data, passed by update form into update mutation.
We need explicitly normalize data before saving it to database.

### CODE REVIEW

File: apps/condo/domains/common/schema/fields.js [hunk 1]
```diff
@@ -41,6 +44,8 @@ const SENDER_FIELD = {
 const ADDRESS_META_FIELD = {
     schemaDoc: 'Property address components',
     type: Json,
+    extendGraphQLTypes: [ADDRESS_META_GRAPHQL_TYPES],
+    graphQLReturnType: 'AddressMeta',
```
Comment (pahaz) [inline apps/condo/domains/common/schema/fields.js:48]: AddressMetaField

File: apps/condo/domains/common/utils/addressApi/addressMetaGql.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
+const ADDRESS_META_GRAPHQL_TYPES = `
```
Comment (pahaz) [inline apps/condo/domains/common/utils/addressApi/addressMetaGql.js:1]: Please move it inside property domain (not common)

File: apps/condo/domains/common/schema/fields.js [hunk 3]
```diff
@@ -2,6 +2,9 @@ const { hasValidJsonStructure } = require('@condo/domains/common/utils/validatio
 const { Integer } = require('@keystonejs/fields')
 const { Json } = require('@core/keystone/fields')
 const { JSON_UNKNOWN_VERSION_ERROR, REQUIRED_NO_VALUE_ERROR, JSON_EXPECT_OBJECT_ERROR } = require('@condo/domains/common/constants/errors')
+const { ADDRESS_META_GRAPHQL_TYPES } = require('../utils/addressApi/addressMetaGql')
+const { normalizeEmail } = require('@condo/domains/common/utils/mail')
```
Comment (pahaz) [inline apps/condo/domains/common/schema/fields.js:6]: ?

File: apps/condo/domains/property/gql.js [hunk 4]
```diff
@@ -5,13 +5,61 @@
...
 const Property = generateGqlQueries('Property', PROPERTY_FIELDS)
 
+const PROPERTY_MAP_GRAPHQL_TYPES = `
+    scalar building
+    scalar section
+    scalar floor
+    scalar unit
+    scalar village
+
+    enum BuildingMapEntity {
```
Comment (pahaz) [inline apps/condo/domains/property/gql.js:23]: BuildingMapEntityType

File: apps/condo/domains/property/gql.js [hunk 5]
```diff
@@ -5,13 +5,61 @@
...
 
 const COMMON_FIELDS = 'id dv sender { dv fingerprint } v deletedAt organization { id name} newId createdBy { id name } updatedBy { id name } createdAt updatedAt'
-
-const PROPERTY_FIELDS = `{ name address addressMeta type ticketsInWork ticketsClosed unitsCount map ${COMMON_FIELDS} }`
+const PROPERTY_MAP_JSON_FIELDS = 'dv type sections { id type index name preview floors { id type index name units { id type name label preview } } }'
+const PROPERTY_FIELDS = `{ name address addressMeta { ${ADDRESS_META_SUBFIELDS_QUERY_LIST} } type ticketsInWork ticketsClosed unitsCount map { ${PROPERTY_MAP_JSON_FIELDS} } ${COMMON_FIELDS} }`
 const Property = generateGqlQueries('Property', PROPERTY_FIELDS)
 
+const PROPERTY_MAP_GRAPHQL_TYPES = `
+    scalar building
```
Comment (pahaz) [inline apps/condo/domains/property/gql.js:17]: we don't need scalar! All our API have only two scalars: JSON and Upload :)

File: apps/condo/domains/property/gql.js [hunk 6]
```diff
@@ -5,13 +5,61 @@
...
+        building
+        section
+        floor
+        unit
+        village
+    }
+
+    type BuildingUnit {
+        id: String!
+        type: unit!
```
Comment (pahaz) [inline apps/condo/domains/property/gql.js:33]: BuildingMapEntityType

File: apps/condo/domains/property/gql.js [hunk 7]
```diff
@@ -5,13 +5,61 @@
...
+        name: String
+        label: String!
+        preview: Boolean
```
Comment (pahaz) [inline apps/condo/domains/property/gql.js:36]: is it really needed?

File: apps/condo/domains/property/gql.js [hunk 8]
```diff
@@ -5,13 +5,61 @@
...
+    enum BuildingMapEntity {
+        building
+        section
+        floor
+        unit
+        village
+    }
+
+    type BuildingUnit {
+        id: String!
```
Comment (pahaz) [inline apps/condo/domains/property/gql.js:32]: @deprecated

File: apps/condo/domains/property/gql.js [hunk 9]
```diff
@@ -5,13 +5,61 @@
...
+        type: unit!
+        name: String
+        label: String!
+        preview: Boolean
+    }
+
+    type BuildingFloor {
+        id: String!
+        type: floor!
```
Comment (pahaz) [inline apps/condo/domains/property/gql.js:41]: BuildingMapEntityType

File: apps/condo/domains/property/gql.js [hunk 10]
```diff
@@ -5,13 +5,61 @@
...
+        index: Int!
+        name: String!
+        units: [BuildingUnit]!
+    }
+
+    type BuildingSection {
+        id: String!
+        type: section!
```
Comment (pahaz) [inline apps/condo/domains/property/gql.js:49]: BuildingMapEntityType

**Comment by pahaz** [comment]
We also need RegisterResidentService

File: apps/condo/domains/common/utils/addressApi/addressMetaGql.js (hunk 2 above)
Comment (AntonAL) [reply apps/condo/domains/common/utils/addressApi/addressMetaGql.js:1]: It is being used by 'Property' and 'Resident' — two different domains.
What the reason to move it to property domain?

File: apps/condo/domains/common/schema/fields.js (hunk 3 above)
Comment (AntonAL) [reply apps/condo/domains/common/schema/fields.js]: Redundant imports

File: apps/condo/domains/property/gql.js (hunk 5 above)
Comment (AntonAL) [reply apps/condo/domains/property/gql.js:17]: Show me, please, how can I ensure `type` fields for `sections`, `floors` and `units` to have constant values for following query:

```graphql
{
  allProperties {
    map {
      type
      sections {
        type
        floors {
          type
          units {
            type
          }
        }
      }
    }
  }
}
```

Now it correctly responds with following:

```json
{
  "data": {
    "allProperties": [
      {
        "map": {
          "type": "building",
          "sections": [
            {
              "type": "section",
              "floors": [
                {
                  "type": "floor",
                  "units": [
                    {
                      "type": "unit"
                    },
                  ]
                }
              ]
            }
          ]
        }
      }
    ]
  }
}
```

File: apps/condo/domains/property/gql.js (hunk 8 above)
Comment (AntonAL) [reply apps/condo/domains/property/gql.js:32]: @dkoviazin ?

File: apps/condo/domains/property/gql.js (hunk 7 above)
Comment (AntonAL) [reply apps/condo/domains/property/gql.js:36]: @dkoviazin ?

File: apps/condo/domains/property/gql.js (hunk 9 above)
Comment (AntonAL) [reply apps/condo/domains/property/gql.js:41]: It must have only one value — "floor", this is the only way, I've found, to ensure this exact value.
When I will use enum type, any value from enum can slip through GraphQL input validation.

File: apps/condo/domains/property/gql.js (hunk 10 above)
Comment (AntonAL) [reply apps/condo/domains/property/gql.js:49]: It must have only one value — "section", this is the only way, I've found, to ensure this exact value.
When I will use enum type, any value from enum can slip through GraphQL input validation.

File: apps/condo/domains/common/schema/fields.js (hunk 1 above)
Comment (AntonAL) [reply apps/condo/domains/common/schema/fields.js]: Changed in `4aa3a8c97`

File: apps/condo/domains/property/gql.js (hunk 4 above)
Comment (AntonAL) [reply apps/condo/domains/property/gql.js]: Renamed in `a1cb10708`

**Comment by AntonAL** [comment]
> [x] We also need RegisterResidentService.addressMeta (RegisterResidentInput)

Implemented in `9f285b0f`

**Comment by AntonAL** [comment]
@dkoviazin , take a look please at types `BuildingUnit`, `BuildingFloor`, `BuildingSection`.
What fields on them are redundant and will never be used?

Is `type` field needed in all of these types?
For what purpose `BuildingUnit` fields `preview` and `name` was reserved?
For what purpose `BuildingFloor` field `name` was reserved? Is it a custom name of a floor in addition to its index?
What about `BuildingSection` field `preview`?

File: apps/condo/domains/property/gql.js (hunk 6 above)
Comment (AntonAL) [reply apps/condo/domains/property/gql.js]: Changed in `bdc55d31`

File: apps/condo/domains/property/gql.js (hunk 10 above)
Comment (AntonAL) [reply apps/condo/domains/property/gql.js]: Changed in `bdc55d31`

File: apps/condo/domains/common/utils/addressApi/addressMetaGql.js (hunk 2 above)
Comment (AntonAL) [reply apps/condo/domains/common/utils/addressApi/addressMetaGql.js]: Moved

**Review by pahaz — APPROVED** [review]

File: apps/condo/domains/common/schema/fields.js [hunk 11]
```diff
@@ -2,6 +2,7 @@ const { hasValidJsonStructure } = require('@condo/domains/common/utils/validatio
 const { Integer } = require('@keystonejs/fields')
 const { Json } = require('@core/keystone/fields')
 const { JSON_UNKNOWN_VERSION_ERROR, REQUIRED_NO_VALUE_ERROR, JSON_EXPECT_OBJECT_ERROR } = require('@condo/domains/common/constants/errors')
+const { ADDRESS_META_FIELD_GRAPHQL_TYPES } = require('../../property/schema/fields/AddressMetaField')
```
Comment (SavelevMatthew) [inline apps/condo/domains/common/schema/fields.js:5]: cross-domain imports should start with `@condo`

File: apps/condo/domains/property/utils/testSchema/factories.js [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    emptyData.geo_lat = faker.address.latitude()
+    emptyData.geo_lon = faker.address.longitude()
+
+    const fullHouseName = [emptyData.house_type, emptyData.house, emptyData.block_type, emptyData.block]
+        .filter(Boolean)
+        .join(' ')
+    const flat = emptyData.flat ? `${emptyData.flat_type} ${emptyData.flat}` : null
+    const value = [emptyData.region_with_type, emptyData.city_with_type, emptyData.street_with_type, fullHouseName, flat]
```
Comment (SavelevMatthew) [inline apps/condo/domains/property/utils/testSchema/factories.js:72]: There're cases when region is not included in value, such as "Свердловская область", even if region is specified,
You should generate fake address with this scenario as well

**Review by SavelevMatthew — APPROVED** [review]
LGTM, but fix fake addressMeta region inclusion in value

**Review by dkoviazin — APPROVED** [review]
Preview is not needed

**Review by dkoviazin — CHANGES_REQUESTED** [review]
map and address meta cannot be updated
```bash
Invalid JSON for property:map [
  '{\n' +
    '  "instancePath": "/sections/0/floors/0/units/0/name",\n' +
    '  "schemaPath": "#/properties/sections/items/properties/floors/items/properties/units/items/properties/name/type",\n' +
    '  "keyword": "type",\n' +
    '  "params": {\n' +
    '    "type": "string"\n' +
    '  },\n' +
    '  "message": "must be string"\n' +
    '}'
]
Invalid JSON for property:map [
  '{\n' +
    '  "instancePath": "/sections/0/floors/0/units/0/name",\n' +
    '  "schemaPath": "#/properties/sections/items/properties/floors/items/properties/units/items/properties/name/type",\n' +
    '  "keyword": "type",\n' +
    '  "params": {\n' +
    '    "type": "string"\n' +
    '  },\n' +
    '  "message": "must be string"\n' +
    '}'
]
```

File: apps/condo/domains/common/schema/fields.js (hunk 11 above)
Comment (AntonAL) [reply apps/condo/domains/common/schema/fields.js]: Fixed in `34b72266`

File: apps/condo/domains/property/utils/testSchema/factories.js (hunk 12 above)
Comment (AntonAL) [reply apps/condo/domains/property/utils/testSchema/factories.js:72]: I think, there is no property on the whole "Свердловская область", it should have some lower-level classification.
We have no properties with blank `addressMeta.data.region_with_type`, tested on prod.
Fields `flat`, `flat_type`, `city_with_type` and `street_with_type` are not required.

I've changed set of required fields after my research of production data in `91242a4b`.

**Comment by AntonAL** [comment]
> To reproduce - add new section to map and try to save
> 
> ```shell
> Invalid JSON for property:map [
>   '{\n' +
>     '  "instancePath": "/sections/0/floors/0/units/0/name",\n' +
>     '  "schemaPath": "#/properties/sections/items/properties/floors/items/properties/units/items/properties/name/type",\n' +
>     '  "keyword": "type",\n' +
>     '  "params": {\n' +
>     '    "type": "string"\n' +
>     '  },\n' +
>     '  "message": "must be string"\n' +
>     '}'
> ]
> Invalid JSON for property:map [
>   '{\n' +
>     '  "instancePath": "/sections/0/floors/0/units/0/name",\n' +
>     '  "schemaPath": "#/properties/sections/items/properties/floors/items/properties/units/items/properties/name/type",\n' +
>     '  "keyword": "type",\n' +
>     '  "params": {\n' +
>     '    "type": "string"\n' +
>     '  },\n' +
>     '  "message": "must be string"\n' +
>     '}'
> ]
> ```

It seems, like raw GraphQL request result is passed into `updateProperty` mutation input, because in mutation variables I see `__typename` fields in `BuildingSection`, `BuildingFloor`, and `BuildingUnit` nodes.

**Review by dkoviazin — APPROVED** [review]

**Pull request merged**